lang_GENERIC/analyze/Dataflow_constness.ml

```diff
@@ -0,0 +1,247 @@
+(*
+ * Copyright (C) 2020 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file license.txt.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * license.txt for more details.
+*)
+open Common
+open IL
+
+module G = AST_generic
+module F = IL
+module D = Dataflow
+module VarMap = Dataflow.VarMap
+
+(*****************************************************************************)
+(* Types *)
+(*****************************************************************************)
+
+type kind =
+  | Kbool
+  | Kint
+  | Kstr
+  | Kany
+
+type constness = Lit of AST_generic.literal | Cst of kind | Dyn
+
+(* map for each node/var whether a variable is "tainted" *)
+type mapping = constness Dataflow.mapping
+
+module DataflowX = Dataflow.Make (struct
+    type node = F.node
+    type edge = F.edge
+    type flow = (node, edge) Ograph_extended.ograph_mutable
+    let short_string_of_node n =
+      Display_IL.short_string_of_node_kind n.F.n
+  end)
+
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+let string_of_kind = function
+  | Kbool -> "bool"
+  | Kint  -> "int"
+  | Kstr  -> "str"
+  | Kany  -> "???"
+
+let string_of_constness = function
+  | Dyn   -> "dyn"
+  | Cst k -> Printf.sprintf "cst(%s)" (string_of_kind k)
+  | Lit l ->
+      match l with
+      | G.Bool (b, _) -> Printf.sprintf "lit(%b)" b
+      | G.Int (s, _)
+      | G.String (s, _) -> Printf.sprintf "lit(%s)" s
+      | ___else___ -> "lit(???)"
+
+let str_of_name ((s, _tok), sid) =
+  spf "%s:%d" s sid
+
+(*****************************************************************************)
+(* Constness *)
+(*****************************************************************************)
+
+(* TODO: Check if this function already exists somewhere else... *)
```

Comment:
nope, we don't have one. In semgrep I abuse the = of OCaml to compare big trees; I just
make sure first to "abstract" the line information attached to each node. See Matching_generic.equal_ast_binded_code.